### ISSUE TITLE
Error using SSL connection with v9.1.17.0 and openJdk 11

### ISSUE THREAD

**Issue by abelsromero** [jruby/jruby#5374]
**Quick summary:** AsciidoctorJ does not work with Java 11. JRuby 9.2.0.0 is not compatible, and the issue in v9.1.17.0 comes of the old version of jopenssl included, manually replacing it fixes it.
Does it makes sense to release a new jruby-complete 9.1.x or should we start thinking about adapting to 9.2.0.0.

### Environment

Provide at least:
* JRuby version (`jruby -v`) and command line (flags, JRUBY_OPTS, etc)
Using jruby-complete 9.1.17.0 to load a gem (Asciidoctor)

* Operating system and platform (e.g. `uname -a`)
Windows and Linux

Other relevant info you may wish to add:
* Installed or activated gems
The main gem is provided in a Jar (AsciidoctorJ).

* Application/framework version (e.g. Rails, Sinatra)
N/A
* Environment variables
N/A

### Expected Behavior

* Describe your expectation of how JRuby should behave, perhaps by showing how CRuby/MRI behaves.
Code should run as normal same as with Java version 8, 9 & 10.

* Provide an executable Ruby script or a link to an example repository.
Reproduce running `mvn clean test` from master of https://github.com/asciidoctor/asciidoctor-maven-plugin. Test `github files can be included` fails.
Also, run tests ('./gradlew clean test') in AsciidoctorJ -> https://github.com/asciidoctor/asciidoctorj/issues/696#issuecomment-430417557.
This tests uses en include remote file from Asciidoctor to pull content from an https resource.

### Actual Behavior

* Describe or show the actual behavior.
Process fails with the message.

```
ERROR:  Loading command: install (LoadError)
        load error: jopenssl/load -- java.lang.StringIndexOutOfBoundsException: begin 0, end 3, length 2
ERROR:  While executing gem ... (NoMethodError)
    undefined method `invoke_with_build_args' for nil:NilClass
```
I could fix this manually installing and requiring the latests `jopenssl` which fixes the issue.

**Comment by mkristian** [comment]
@abelsromero from your pom I see only jruby-complete and no jruby-maven-plugin or mavengem extension. how to 'get' the asciidoctor gem ? I can not see it.

**Comment by abelsromero** [comment]
Apologies, but you'll have to go slower here. I did see the other issue, but I don't understanding what you mean and imho these are not the same. Please correct if I am wrong.

I extracted jruby-complete and Asciidoctorj jars to look for openssl and I only found the folder in jruby-complete, so I understand there's no repackage done on Asciidoctorj side. My test project has no other dependencies and only uses gems from these jars.
When checking jruby-complete 9.1.17.0, I see this the path `jruby-complete-9.1.17.0\META-INF\jruby.home\lib\ruby\stdlib\jopenssl\version.rb` 

```ruby
module Jopenssl
  VERSION = '0.9.21'
  BOUNCY_CASTLE_VERSION = '1.56'
  # @deprecated
  module Version
    # @private
    VERSION = Jopenssl::VERSION
    # @private
    BOUNCY_CASTLE_VERSION = Jopenssl::BOUNCY_CASTLE_VERSION
  end
end
```

version 9.2.0.0 contains

```ruby
module Jopenssl
  VERSION = '0.10.0'
  BOUNCY_CASTLE_VERSION = '1.59'
end
```


From that, I understand that jruby-complete contains an old openssl version that has the issue, and that to fix AsciidoctorJ I have 3 options:
1. Use a new jruby-complete 9.1.18.0 version -> that I leave for you to decide whether it makes sense or not to release.
2. Adapt AsciidoctorJ to use 9.2.0.0
3. Current workarround, install a newer version of the gem using gem-maven/gradle plugin and use a feature provided by AsciidoctorJ that allows to override the one in the jar -> This is very cumbersome and not really a good solution users not familiar with Ruby, the goal of AsciidoctorJ is to hide this details.

Am I wrong and the `version.rb` I am seeing has nothing to do with the version?

**Comment by mkristian** [comment]
@abelsromero thanks that new pair of eyes did help indeed.

@kares I just doublechecked jruby-complete-9.1.17.0.jar from maven central comes with jossl-0.9.21 both ruby code as well jossl.jar

this finally explains our mystery why we still saw 0.9.21 errors though we all thought jruby-9.1.17.0 uses jossl-0.10.0 !

did not check jruby bin-distribution

cc @headius @enebo

**Comment by kares** [comment]
hmm, you're right ... I might have a modified dist of 9.1.17 with 0.10 - the 9.1 branch still has 0.9.21 which contains some code around fixing version num parsing but 0.10 has more. anyway, at this point we can not fix it for 9.1.17 and mosy effort goes through 9.2.x ... maybe we shall just declare 9.1 not supporting Java 11. altough mvn plugin with 9.1 does boot on 11, right?

**Comment by abelsromero** [comment]
Being selfish here. I'd like to provide java11 in Asciidoctorj right now. So, if a minor release 9.1.17.x does not require a big effort it would be a really nice gesture toward the project and also others that may encounter similar issues.

Now, being aware that 9.1.17.x won't get any more support, we can already start on migrating, but I suspect won't be easy.

**Comment by headius** [comment]
@abelsromero What do you expect to be the problems migrating to JRuby 9.2?

The 9.1.x line could do with one more release (cc @enebo) but if we can avoid that it would be best. Of course we have not done a maintenance release of 9.2 either.

**Comment by abelsromero** [comment]
> @abelsromero What do you expect to be the problems migrating to JRuby


I assume 9.2.0.0 is not backward compatible given the number change and the errors I get when updating both `Asciidoctorj` and `asciidoctor-maven-plugin` to 9.2.0.0.
The error makes me thing that there's a change how jRuby represents the internal Ruby objects.

```
java.lang.VerifyError: Bad return type
Exception Details:
  Location:
    org/jruby/gen/InterfaceImpl866559659.resolve_converter(Ljava/lang/String;)Lorg/jruby/RubyClass; @54: areturn
  Reason:
    Type 'org/jruby/runtime/builtin/IRubyObject' (current frame, stack[0]) is not assignable to 'org/jruby/RubyClass' (from method signature)
  Current Frame:
    bci: @54
    flags: { }
    locals: { 'org/jruby/gen/InterfaceImpl866559659', 'java/lang/String', 'org/jruby/runtime/builtin/IRubyObject', 'org/jruby/Ruby' }
    stack: { 'org/jruby/runtime/builtin/IRubyObject' }
  Bytecode:
    0x0000000: 2ab4 0015 4d2c b900 1d01 004e b200 1f2c
    0x0000010: 12c5 12c6 b600 272d b600 2d2c 2cb9 0031
    0x0000020: 0100 12c6 04bd 0019 5903 2d2b b800 3753
    0x0000030: b200 3db6 0043 b0                      

	at java.lang.Class.getDeclaredConstructors0(Native Method)
	at java.lang.Class.privateGetDeclaredConstructors(Class.java:2671)
	at java.lang.Class.getConstructor0(Class.java:3075)
	at java.lang.Class.getConstructor(Class.java:1825)
	at org.jruby.javasupport.Java.newInterfaceImpl(Java.java:1352)
	at org.jruby.java.proxies.JavaInterfaceTemplate.newInterfaceProxy(JavaInterfaceTemplate.java:285)
	at org.jruby.java.proxies.JavaInterfaceTemplate.access$100(JavaInterfaceTemplate.java:34)
	at org.jruby.java.proxies.JavaInterfaceTemplate$InterfaceProxyFactory.call(JavaInterfaceTemplate.java:233)
	at org.jruby.internal.runtime.methods.DynamicMethod.call(DynamicMethod.java:194)
	at org.jruby.RubyClass.finvoke(RubyClass.java:560)
	at org.jruby.runtime.Helpers.invoke(Helpers.java:353)
	at org.jruby.javasupport.JavaUtil.convertProcToInterface(JavaUtil.java:278)
	at org.jruby.RubyBasicObject.defaultToJava(RubyBasicObject.java:898)
	at org.jruby.RubyBasicObject.toJava(RubyBasicObject.java:876)
	at org.jruby.javasupport.JavaEmbedUtils.rubyToJava(JavaEmbedUtils.java:247)
	at org.asciidoctor.internal.RubyUtils.rubyToJava(RubyUtils.java:19)
```

**Comment by enebo** [comment]
I have been merging some stuff back to 9.1 on occasion so I think we can definitely do another 9.1.x release.

Regardless of changes between 9.1 -> 9.2 you should never see a class verification error.  This is flat out a bug.  It may have been fixed since 9.2.0.0.  Trying 9.2.1.0-SNAPSHOT would be worth a try and help us to know if we still have something to fix.  That is definitely going to be released in next couple of weeks (as has been said several times in the past but rubyconf is coming so we for sure will have it out for rubyconf).

**Comment by abelsromero** [comment]
Good point @enebo!
AsciidoctorJ and maven plugin work fine with 9.2.1.0-SNAPSHOT (build from master) with Java 8 , 9 and 11. I think we can wait a few weeks for the official release.

PS: should the `Invalid or Duplicate` milestone be removed?

**Comment by headius** [comment]
Please file the verification error separately if you still see it. @enebo is right, that definitely should not be happening.

If there's nothing to do for this it could be marked for 9.2.1 or just left as invalid (since the original report is already covered by other fixed issues).

**Comment by abelsromero** [comment]
> If there's nothing to do for this it could be marked for 9.2.1 or just left as invalid (since the original report is already covered by other fixed issues).

Closing and marking for 9.2.1 is fine for me. Asciidoctor release won't be delayed by this.
As soon as the release is available I'll test and open a new issue if it still happens 😉 

thx to all!

**Comment by kares** [comment]
seems like this was resolved either in 9.1.18.0 or 9.2.1.0
